RetryingClient does not have a child request if a root request is already done

I'm pretty not sure how to reproduce it but here is the stacktrace from our service.

* OS: Linux
* Client: Armeria-0.91.0
* Server: Armeria-0.88.0

```
Caused by: java.lang.IllegalStateException: at least one child should be already added
	at com.linecorp.armeria.internal.shaded.guava.base.Preconditions.checkState(Preconditions.java:508)
	at com.linecorp.armeria.common.logging.DefaultRequestLog.endResponseWithLastChild(DefaultRequestLog.java:234)
	at com.linecorp.armeria.client.retry.RetryingClient.onRetryingComplete(RetryingClient.java:128)
	at com.linecorp.armeria.client.retry.RetryingHttpClient.handleException(RetryingHttpClient.java:204)
	at com.linecorp.armeria.client.retry.RetryingHttpClient.doExecute0(RetryingHttpClient.java:151)
	at com.linecorp.armeria.client.retry.RetryingHttpClient.doExecute(RetryingHttpClient.java:141)
	at com.linecorp.armeria.client.retry.RetryingHttpClient.doExecute(RetryingHttpClient.java:58)
	at com.linecorp.armeria.client.retry.RetryingClient.execute(RetryingClient.java:115)
	at com.linecorp.armeria.client.metric.MetricCollectingClient.execute(MetricCollectingClient.java:73)
	at com.linecorp.armeria.client.tracing.HttpTracingClient.execute(HttpTracingClient.java:154)
	at com.linecorp.armeria.client.tracing.HttpTracingClient.execute(HttpTracingClient.java:63)
	at com.linecorp.armeria.internal.ClientUtil.pushAndExecute(ClientUtil.java:82)
	at com.linecorp.armeria.internal.ClientUtil.initContextAndExecuteWithFallback(ClientUtil.java:51)
	at com.linecorp.armeria.client.UserClient.execute(UserClient.java:148)
	at com.linecorp.armeria.client.DefaultHttpClient.execute(DefaultHttpClient.java:99)
	at com.linecorp.armeria.client.DefaultHttpClient.execute(DefaultHttpClient.java:90)
	at com.linecorp.armeria.client.DefaultHttpClient.execute(DefaultHttpClient.java:117)
	at com.linecorp.armeria.client.DefaultHttpClient.execute(DefaultHttpClient.java:113)
	at com.linecorp.armeria.client.HttpClient.execute(HttpClient.java:217)
	at com.linecorp.armeria.client.HttpClient.get(HttpClient.java:259)
```

--- a/core/src/main/java/com/linecorp/armeria/client/retry/RetryingClient.java
+++ b/core/src/main/java/com/linecorp/armeria/client/retry/RetryingClient.java
@@ -257,9 +257,9 @@ public abstract class RetryingClient<I extends Request, O extends Response>
      * If {@link ClientRequestContext#endpointSelector()} exists, a new {@link Endpoint} will be selected.
      */
     protected static ClientRequestContext newDerivedContext(ClientRequestContext ctx,
-                                                            Request req, int totalAttempts) {
+                                                            Request req, boolean initialAttempt) {
         final EndpointSelector endpointSelector = ctx.endpointSelector();
-        if (endpointSelector != null && totalAttempts > 1) {
+        if (endpointSelector != null && !initialAttempt) {
             return ctx.newDerivedContext(req, endpointSelector.select(ctx));
         } else {
             return ctx.newDerivedContext(req);

--- a/core/src/main/java/com/linecorp/armeria/client/retry/RetryingHttpClient.java
+++ b/core/src/main/java/com/linecorp/armeria/client/retry/RetryingHttpClient.java
@@ -145,21 +145,22 @@ public final class RetryingHttpClient extends RetryingClient<HttpRequest, HttpRe
     private void doExecute0(ClientRequestContext ctx, HttpRequestDuplicator rootReqDuplicator,
                             HttpRequest originalReq, HttpResponse returnedRes,
                             CompletableFuture<HttpResponse> future, boolean hasInitialAuthority) {
+        final int totalAttempts = getTotalAttempts(ctx);
+        final boolean initialAttempt = totalAttempts <= 1;
         if (originalReq.completionFuture().isCompletedExceptionally() || returnedRes.isComplete()) {
             // The request or response has been aborted by the client before it receives a response,
             // so stop retrying.
-            handleException(ctx, rootReqDuplicator, future, AbortedStreamException.get());
+            handleException(ctx, rootReqDuplicator, future, AbortedStreamException.get(), initialAttempt);
             return;
         }
 
         if (!setResponseTimeout(ctx)) {
-            handleException(ctx, rootReqDuplicator, future, ResponseTimeoutException.get());
+            handleException(ctx, rootReqDuplicator, future, ResponseTimeoutException.get(), initialAttempt);
             return;
         }
 
-        final int totalAttempts = getTotalAttempts(ctx);
         final HttpRequest duplicateReq;
-        if (hasInitialAuthority && totalAttempts <= 1) {
+        if (hasInitialAuthority && initialAttempt) {
             duplicateReq = rootReqDuplicator.duplicateStream();
         } else {
             final RequestHeadersBuilder newHeaders = originalReq.headers().toBuilder();
@@ -172,7 +173,7 @@ public final class RetryingHttpClient extends RetryingClient<HttpRequest, HttpRe
             duplicateReq = rootReqDuplicator.duplicateStream(newHeaders.build());
         }
 
-        final ClientRequestContext derivedCtx = newDerivedContext(ctx, duplicateReq, totalAttempts);
+        final ClientRequestContext derivedCtx = newDerivedContext(ctx, duplicateReq, initialAttempt);
         ctx.logBuilder().addChild(derivedCtx.log());
 
         final HttpResponse response = executeWithFallback(delegate(), derivedCtx,
@@ -200,8 +201,12 @@ public final class RetryingHttpClient extends RetryingClient<HttpRequest, HttpRe
     }
 
     private static void handleException(ClientRequestContext ctx, HttpRequestDuplicator rootReqDuplicator,
-                                        CompletableFuture<HttpResponse> future, Throwable cause) {
-        onRetryingComplete(ctx);
+                                        CompletableFuture<HttpResponse> future, Throwable cause,
+                                        boolean endRequestLog) {
+        if (endRequestLog) {
+            ctx.logBuilder().endRequest(cause);
+        }
+        ctx.logBuilder().endResponse(cause);
         future.completeExceptionally(cause);
         rootReqDuplicator.close();
     }
@@ -233,7 +238,7 @@ public final class RetryingHttpClient extends RetryingClient<HttpRequest, HttpRe
                 if (nextDelay >= 0) {
                     closingOriginalResTask.run();
                     scheduleNextRetry(
-                            ctx, cause -> handleException(ctx, rootReqDuplicator, future, cause),
+                            ctx, cause -> handleException(ctx, rootReqDuplicator, future, cause, false),
                             () -> doExecute0(ctx, rootReqDuplicator, originalReq,
                                              returnedRes, future, hasInitialAuthority),
                             nextDelay);

--- a/core/src/main/java/com/linecorp/armeria/client/retry/RetryingRpcClient.java
+++ b/core/src/main/java/com/linecorp/armeria/client/retry/RetryingRpcClient.java
@@ -90,22 +90,23 @@ public final class RetryingRpcClient extends RetryingClient<RpcRequest, RpcRespo
 
     private void doExecute0(ClientRequestContext ctx, RpcRequest req,
                             RpcResponse returnedRes, CompletableFuture<RpcResponse> future) {
+        final int totalAttempts = getTotalAttempts(ctx);
+        final boolean initialAttempt = totalAttempts <= 1;
         if (returnedRes.isDone()) {
             // The response has been cancelled by the client before it receives a response, so stop retrying.
             handleException(ctx, future, new CancellationException(
-                    "the response returned to the client has been cancelled"));
+                    "the response returned to the client has been cancelled"), initialAttempt);
             return;
         }
         if (!setResponseTimeout(ctx)) {
-            handleException(ctx, future, ResponseTimeoutException.get());
+            handleException(ctx, future, ResponseTimeoutException.get(), initialAttempt);
             return;
         }
 
-        final int totalAttempts = getTotalAttempts(ctx);
-        final ClientRequestContext derivedCtx = newDerivedContext(ctx, req, totalAttempts);
+        final ClientRequestContext derivedCtx = newDerivedContext(ctx, req, initialAttempt);
         ctx.logBuilder().addChild(derivedCtx.log());
 
-        if (totalAttempts > 1) {
+        if (!initialAttempt) {
             derivedCtx.setAdditionalRequestHeader(ARMERIA_RETRY_COUNT, Integer.toString(totalAttempts - 1));
         }
 
@@ -122,7 +123,7 @@ public final class RetryingRpcClient extends RetryingClient<RpcRequest, RpcRespo
                         return null;
                     }
 
-                    scheduleNextRetry(ctx, cause -> handleException(ctx, future, cause),
+                    scheduleNextRetry(ctx, cause -> handleException(ctx, future, cause, false),
                                       () -> doExecute0(ctx, req, returnedRes, future), nextDelay);
                 } else {
                     onRetryingComplete(ctx);
@@ -135,8 +136,11 @@ public final class RetryingRpcClient extends RetryingClient<RpcRequest, RpcRespo
     }
 
     private static void handleException(ClientRequestContext ctx, CompletableFuture<RpcResponse> future,
-                                        Throwable cause) {
-        onRetryingComplete(ctx);
+                                        Throwable cause, boolean endRequestLog) {
+        if (endRequestLog) {
+            ctx.logBuilder().endRequest(cause);
+        }
+        ctx.logBuilder().endResponse(cause);
         future.completeExceptionally(cause);
     }
 }

--- a/core/src/test/java/com/linecorp/armeria/client/retry/RetryingHttpClientTest.java
+++ b/core/src/test/java/com/linecorp/armeria/client/retry/RetryingHttpClientTest.java
@@ -53,6 +53,7 @@ import com.linecorp.armeria.client.HttpClient;
 import com.linecorp.armeria.client.HttpClientBuilder;
 import com.linecorp.armeria.client.ResponseTimeoutException;
 import com.linecorp.armeria.client.UnprocessedRequestException;
+import com.linecorp.armeria.client.logging.LoggingClient;
 import com.linecorp.armeria.common.AggregatedHttpResponse;
 import com.linecorp.armeria.common.HttpData;
 import com.linecorp.armeria.common.HttpHeaderNames;
@@ -64,6 +65,9 @@ import com.linecorp.armeria.common.HttpResponse;
 import com.linecorp.armeria.common.HttpStatus;
 import com.linecorp.armeria.common.MediaType;
 import com.linecorp.armeria.common.ResponseHeaders;
+import com.linecorp.armeria.common.logging.RequestLog;
+import com.linecorp.armeria.common.logging.RequestLogAvailability;
+import com.linecorp.armeria.common.stream.AbortedStreamException;
 import com.linecorp.armeria.common.util.EventLoopGroups;
 import com.linecorp.armeria.server.AbstractHttpService;
 import com.linecorp.armeria.server.ServerBuilder;
@@ -417,6 +421,7 @@ public class RetryingHttpClientTest {
                 .decorator(new RetryingHttpClientBuilder(
                         // Retry after 8000 which is slightly less than responseTimeoutMillis(10000).
                         RetryStrategy.onServerErrorStatus(Backoff.fixed(8000))).newDecorator())
+                .decorator(LoggingClient.newDecorator())
                 .build();
 
         // There's no way to notice that the RetryingClient has scheduled the next retry.
@@ -447,11 +452,23 @@ public class RetryingHttpClientTest {
 
     @Test
     public void doNotRetryWhenResponseIsAborted() throws Exception {
-        final HttpClient client = client(retryAlways);
+        final AtomicReference<ClientRequestContext> context = new AtomicReference<>();
+        final HttpClient client = new HttpClientBuilder(server.uri("/"))
+                .decorator(RetryingHttpClient.newDecorator(retryAlways))
+                .decorator((delegate, ctx, req) -> {
+                    context.set(ctx);
+                    return delegate.execute(ctx, req);
+                })
+                .build();
         final HttpResponse httpResponse = client.get("/response-abort");
         httpResponse.abort();
 
-        await().untilAsserted(() -> assertThat(responseAbortServiceCallCounter.get()).isOne());
+        final RequestLog log = context.get().log();
+        await().untilAsserted(() -> assertThat(log.isAvailable(RequestLogAvailability.COMPLETE)).isTrue());
+        assertThat(responseAbortServiceCallCounter.get()).isOne();
+        assertThat(log.requestCause()).isNull();
+        assertThat(log.responseCause()).isExactlyInstanceOf(AbortedStreamException.class);
+
         // Sleep 3 seconds more to check if there was another retry.
         TimeUnit.SECONDS.sleep(3);
         assertThat(responseAbortServiceCallCounter.get()).isOne();
@@ -482,7 +499,14 @@ public class RetryingHttpClientTest {
 
     @Test
     public void doNotRetryWhenRequestIsAborted() throws Exception {
-        final HttpClient client = client(retryAlways);
+        final AtomicReference<ClientRequestContext> context = new AtomicReference<>();
+        final HttpClient client = new HttpClientBuilder(server.uri("/"))
+                .decorator(RetryingHttpClient.newDecorator(retryAlways))
+                .decorator((delegate, ctx, req) -> {
+                    context.set(ctx);
+                    return delegate.execute(ctx, req);
+                })
+                .build();
 
         final HttpRequestWriter req = HttpRequest.streaming(HttpMethod.GET, "/request-abort");
         req.write(HttpData.ofUtf8("I'm going to abort this request"));
@@ -492,6 +516,9 @@ public class RetryingHttpClientTest {
         TimeUnit.SECONDS.sleep(1);
         // No request is made.
         assertThat(responseAbortServiceCallCounter.get()).isZero();
+        final RequestLog log = context.get().log();
+        assertThat(log.requestCause()).isExactlyInstanceOf(AbortedStreamException.class);
+        assertThat(log.responseCause()).isExactlyInstanceOf(AbortedStreamException.class);
     }
 
     @Test

--- a/thrift/src/test/java/com/linecorp/armeria/it/client/retry/RetryingRpcClientTest.java
+++ b/thrift/src/test/java/com/linecorp/armeria/it/client/retry/RetryingRpcClientTest.java
@@ -34,6 +34,7 @@ import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.Executors;
 import java.util.concurrent.TimeUnit;
 import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
 
 import org.apache.thrift.TApplicationException;
 import org.junit.Rule;
@@ -42,12 +43,15 @@ import org.junit.Test;
 import com.linecorp.armeria.client.ClientBuilder;
 import com.linecorp.armeria.client.ClientFactory;
 import com.linecorp.armeria.client.ClientFactoryBuilder;
+import com.linecorp.armeria.client.ClientRequestContext;
 import com.linecorp.armeria.client.UnprocessedRequestException;
 import com.linecorp.armeria.client.retry.Backoff;
 import com.linecorp.armeria.client.retry.RetryStrategyWithContent;
 import com.linecorp.armeria.client.retry.RetryingRpcClient;
 import com.linecorp.armeria.client.retry.RetryingRpcClientBuilder;
 import com.linecorp.armeria.common.RpcResponse;
+import com.linecorp.armeria.common.logging.RequestLog;
+import com.linecorp.armeria.common.logging.RequestLogAvailability;
 import com.linecorp.armeria.common.util.EventLoopGroups;
 import com.linecorp.armeria.server.ServerBuilder;
 import com.linecorp.armeria.server.thrift.THttpService;
@@ -200,9 +204,11 @@ public class RetryingRpcClientTest {
 
     @Test
     public void doNotRetryWhenResponseIsCancelled() throws Exception {
+        final AtomicReference<ClientRequestContext> context = new AtomicReference<>();
         final HelloService.Iface client = new ClientBuilder(server.uri(BINARY, "/thrift"))
                 .rpcDecorator(new RetryingRpcClientBuilder(retryAlways).newDecorator())
                 .rpcDecorator((delegate, ctx, req) -> {
+                    context.set(ctx);
                     final RpcResponse res = delegate.execute(ctx, req);
                     res.cancel(true);
                     return res;
@@ -211,7 +217,12 @@ public class RetryingRpcClientTest {
         when(serviceHandler.hello(anyString())).thenThrow(new IllegalArgumentException());
 
         assertThatThrownBy(() -> client.hello("hello")).isInstanceOf(CancellationException.class);
-        await().untilAsserted(() -> verify(serviceHandler, only()).hello("hello"));
+
+        final RequestLog log = context.get().log();
+        await().untilAsserted(() -> assertThat(log.isAvailable(RequestLogAvailability.COMPLETE)).isTrue());
+        verify(serviceHandler, only()).hello("hello");
+        assertThat(log.requestCause()).isNull();
+        assertThat(log.responseCause()).isExactlyInstanceOf(CancellationException.class);
 
         // Sleep 1 second more to check if there was another retry.
         TimeUnit.SECONDS.sleep(1);

